### ISSUE TITLE
Design recovery mechanisms for node crash

### ISSUE THREAD

**Issue by kansi** [bigchaindb/bigchaindb#2018]
### Description
As of now the transactions are written to the database during `deliver_tx`, with #2017 the transaction will be written during block `commit`. Since we are using mongodb as the underlying database, `insert_many` is not transactional. So, in case `insert_many` crashes midway the database will be left in an inconsistent state. This crash might be because of improper bigchaindb shutdown, system failure etc.

### Solutions
This error could be handeled in various ways,

#### Approach 1
Catch when `insert_many` throws error and delete the documents which were successfully inserted so as to return to a valid state (two-phase-commit). This approach may work in some cases but will fall short in scenarios where the bigchain process is killed for some reason or the vm shuts down abruptly etc.

#### Approach 2
Devise a recovery mechanism i.e. during `commit` the last db operation is to store the block (which contains a list of transaction ids for that block and app hash) after which the db is in a consistent state. After this the new app hash is returned to tendermint. So, when a crash occurs bigchaindb could be doing the following:
 - failed while writing transactions with `insert_many`
 - Wrote transactions but failed to write the block
 - Wrote transaction and block but failed while returning the new app hash

In general what we might have is a bunch of transactions which don't belong to any block or an app hash which is not registered back to tendermint.  To get to a consistent state we can create a command `bigchaindb recover` which would find and delete "dangling" transactions i.e. transactions which are not included in any block. Furthermore, we delete the latest committed block along with its transactions which would handle the case where transactions and block are written successfully but failure occurred while returning the new app hash to tendermint.

To summarize: "dangling" transactions and the latest block are deleted so when Bighchaindb is started back again it will sync up and start adding new transactions.

NOTE: We could implement both approach 1 & 2

**Comment by ttmc** [comment]
Both 1 & 2 seem like good ideas to me. :+1:

**Comment by ttmc** [comment]
@kansi After you created this issue / proposal, MongoDB announced that:

> MongoDB 4.0 will add support for multi-document transactions, making it the only database to combine the speed, flexibility, and power of the document model with ACID guarantees. Through snapshot isolation, transactions will provide a globally consistent view of data, and enforce all-or-nothing execution to maintain data integrity. MongoDB 4.0 will ship in Summer 2018.

Source:  https://www.mongodb.com/transactions

I wonder if we should create a placeholder issue to revisit this "crash recovery mechanism" when we upgrade to MongoDB 4.0? I get the sense that there would be nothing to recover _from_.

**Comment by kansi** [comment]
@ttmc as far as I understand the multi-document transaction is within a collection and not across multiple collections right? If that is the case then it doesn't change a lot of things i.e. we can update the store query to use this feature but the code which recovers from an inconsistant state would still be the same.

**Comment by ttmc** [comment]
@kansi I think this issue can be closed. It seems that GitHub stopped auto-closing issues when pull-request comments contain phrases like `fixes #123456`. I don't know if that's a bug or on purpose.

**Comment by ttmc** [comment]
I think I figured out why GitHub isn't auto-closing issues for us. It only does that if the pull request is merged into your repository's default branch, which for us is `master`. Lately, we've been merging everything into the `tendermint` branch.

https://help.github.com/articles/closing-issues-using-keywords/

**Comment by kansi** [comment]
@ttmc So I am manually closing this issue.

**Comment by vrde** [comment]
I'm reopening this issue since we rolled back some changes. @kansi opened [an issue](https://github.com/tendermint/tendermint/issues/1254) in the Tendermint repo asking for more information that can help us solving this issue.

**Comment by vrde** [comment]
Solved by [8/CRASH-RECOVERY](https://github.com/bigchaindb/BEPs/tree/master/8).
